orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py

```diff
@@ -0,0 +1,731 @@
+        model.eval()
+    else:
+        model.train()
+
+    inputs_on_cuda = [input_.to(device) for input_ in input_list]
+    output = model(*inputs_on_cuda)
+    forward_outputs=[output]
+    grad_outputs = []
+
+    if not is_eval_mode:
+        criterion = torch.nn.MSELoss()
+        target = label_input.to(device)
+        loss = criterion(output, target)
+        loss.backward()
+        torch.cuda.synchronize()
+
+        for name, param in model.named_parameters():
+            if param.requires_grad:
+                grad_outputs.append(param.grad)
+    else:
+        torch.cuda.synchronize()
+    return forward_outputs, grad_outputs
+
+def compare_numpy_list(val_a, val_b):
+    for np_a, np_b in zip(val_a, val_b):
+        equal_ = np.allclose(np_a, np_b, 1e-7, 1e-6, equal_nan=True)
+        if equal_ is False:
+            print("== details ==")
+            k=np_a.reshape(-1)[:100]
+            l=np_b.reshape(-1)[:100]
+            is_equal = np.isclose(k, l, 1e-7, 1e-6, equal_nan=True)
+            res = (is_equal + 1) % 2
+            diff_indices = np.nonzero(res)
+            print(diff_indices)
+            print(k, l)
+            raise ValueError("find a diff")
+
+    print("outputs matched successfully.")
+
+
+def run_training_test_and_compare(pt_model_builder_func, pt_model_inputs_generator, pt_model_label_input, ignore_grad_compare=False):
+    m = pt_model_builder_func()
+    x = pt_model_inputs_generator()
+
+    m_ort = copy.deepcopy(m)
+    x_ort = copy.deepcopy(x)
+
+    outputs, grads = run_with_pytorch_on_gpu(m, [x], pt_model_label_input)
+    torch.cuda.synchronize()
+
+    outputs_ort, grads_ort = run_with_ort_on_gpu(m_ort, [x_ort], pt_model_label_input)
+
+    torch.cuda.synchronize()
+    val_a = [o.detach().cpu().numpy() for o in outputs if o is not None]
+    val_b = [o.detach().cpu().numpy() for o in outputs_ort if o is not None]
+    compare_numpy_list(val_a, val_b)
+
+    # For some test, it is expected the diff might be big due to inconsistent computation orders.
+    if ignore_grad_compare is False:
+        val_a = [o.detach().cpu().numpy() for o in grads if o is not None]
+        val_b = [o.detach().cpu().numpy() for o in grads_ort if o is not None]
+        compare_numpy_list(val_a, val_b)
+
+def run_evalue_test_and_compare(pt_model_builder_func, pt_model_inputs_generator, pt_model_label_input):
+    m = pt_model_builder_func()
+    x = pt_model_inputs_generator()
+
+    m_ort = copy.deepcopy(m)
+    x_ort = copy.deepcopy(x)
+
+    outputs, grads = run_with_pytorch_on_gpu(m, [x], pt_model_label_input, is_eval_mode=True)
+    torch.cuda.synchronize()
+
+    outputs_ort, grads_ort = run_with_ort_on_gpu(m_ort, [x_ort], pt_model_label_input, is_eval_mode=True)
+
+    torch.cuda.synchronize()
+    val_a = [o.detach().cpu().numpy() for o in outputs if o is not None]
+    val_b = [o.detach().cpu().numpy() for o in outputs_ort if o is not None]
+    compare_numpy_list(val_a, val_b)
```

Comment:
how about using _test_helpers.assert_values_are_close() ? 
it uses torch.all_close on torch tensor directly.

Reply:
Yeah, looks reasonable. while I recalled I did not for some reason. Let me refine this.